framework/source/class/qx/test/ui/table/model/Simple.js

```diff
@@ -48,6 +48,84 @@ qx.Class.define("qx.test.ui.table.model.Simple",
       this.assertEquals("test2", data.col2);
 
       tableModel.dispose();
+    },
+    
+    getStringValues : function() {
+      return ["aaaa", "bbbb", "cccc", "dddd", "eeee", "ffff"];
+    },
+    
+    createRandomRows : function(rowCount)
+    {
+      var rowData = [];
+      var strings = this.getStringValues();
+      for (var row = 0; row < rowCount; row++) {
+        rowData.push([ row+1, strings[row % strings.length] ]);
+      }
+      return rowData;
+    },
+    
+    testCustomSortFunction : function()
+    {
+      var stringValues = this.getStringValues();
+      var rowCount = 20;
+      
+      // table
+      var model = new qx.ui.table.model.Simple();
+      model.setColumns([ "ID", "String" ]);
+      model.setData(this.createRandomRows(rowCount));
+
+      // custom individual ascending descending integer sort function for column 0
+      model.setSortMethods(0, {
+        ascending : function(row1, row2, columnIndex) {
+          var int1 = row1[columnIndex];
+          var int2 = row2[columnIndex];
+          return (int1 > int2) ? 1 : ((int1 == int2) ? 0 : -1);
+        },
+        descending : function(row2, row1, columnIndex) {
+          var int1 = row1[columnIndex];
+          var int2 = row2[columnIndex];
+          return (int1 > int2) ? 1 : ((int1 == int2) ? 0 : -1);
+        } 
+      });
+      
+      // custom single string sort function for column 1
+      model.setSortMethods(1, function(row1, row2, columnIndex) {
+        var string1 = row1[columnIndex];
```

Comment:
It just occurred to me... maybe we should `assert(columnIndex === arguments.callee.columnIndex)` here, to prove backwards compatability?

Reply:
You're right. But I think it is better to have a separate test for the `arguments.callee.columnIndex` way.

I've done that here locally and found that it fails for the case that a single comparator function is used and not a map having separate functions for ascending and descending sorting.

I think that this would also have failed for the version before my PR. Please have a look into the unchanged code at https://github.com/qooxdoo/qooxdoo/blob/master/framework/source/class/qx/ui/table/model/Simple.js#L384-L387 
As you can see the ascending function gets wrapped into a function by changing the order of row1 and row2 in the parameters.
Later on when the sort is prepared, the columnIndex attribute to the function object is set here: https://github.com/qooxdoo/qooxdoo/blob/master/framework/source/class/qx/ui/table/model/Simple.js#L328

But this does not work for the formerly in `setSortMethods` wrapped descending function as within the innert function you have no access to `arguments.callee.columnIndex` as this belongs to the outer function.

A fix for the old version would be to set the column index for the inner function in the original version within the wrapper like this:
```javascript
        methods =
          {
            ascending  : compare,
            descending : function(row1, row2)
            {
              compare.columnIndex = arguments.callee.columnIndex;
              return compare(row2, row1);
            }
          };
```
or for the new version it would be:
```javascript
        methods =
          {
            ascending  : compare,
            descending : function(row1, row2, columnIndex)
            {
              compare.columnIndex = arguments.callee.columnIndex;
              return compare(row2, row1, columnIndex);
            }
          };
```
which would lead to having `arguments.callee` again in the framework code.

Or we can argue that having the non working version is backwards compatible to the old one.

Reply:
What a mess. Yes, I see that it's been broken for years and no one has noticed it. Either they're always passing two functions (the better way to do it anyway), or they just haven't noticed.

I'm torn between "fix the bug in the old code" and "the old code was uncomprehendable and fixing the bug affects BC so maybe they should just use the new way if they are going to alter their code anyway"...

Since it's clearly a bug, I think we should fix it, as you've indicated here.

Reply:
Well the new method with the third parameter `columnIndex` fixes it already, but not for the backwards compatibility way via `arguments.callee.columnIndex`.
So you argue to add that line
```javascript
compare.columnIndex = arguments.callee.columnIndex;
```
to the wrapper?
But all this started as an attempt to remove all occurences of `arguments.callee` in the framework code.
I'm undecided what to do.

Reply:
Yes it works the old way if the sort function uses the `arguments.callee.columnIndex` method including having the same bug as it had before in the case of giving a single function to `setSortMethods`.
But it fixes the bug if the new method with the third parameter `columnIndex` is used in the sort functions for all methods either single sort function or map with two separate sort functions for ascending and descending sort.

Reply:
Ok. I think if you add the test you wrote that shows that (possibly with a huge comment saying that we're retaining the existing bug), we'll be good to go.

Reply:
I think I'll add the line this way:
```javascript
// assure backwards compatibility for sort functions using
// arguments.callee.columnIndex and to fix a bug where
// this method did not work for the case where a single comparator
// function was used 
// See discussion in https://github.com/qooxdoo/qooxdoo/pull/9499#pullrequestreview-99655182
if(arguments.callee !== undefined) {
  compare.columnIndex = arguments.callee.columnIndex;
}

Reply:
Really good.

Reply:
AIUI the old code is so broken that if someone was using it, then it would not work?  In which case, then surely BC is not an issue?

Also, you could check and emit a warning, eg 
```
if (qx.core.Environment.get("qx.debug") &&
    compare.toString().match(/arguments.callee.columnIndex/)) {
  this.debug("Warning:: .....");
}
```

Reply:
@johnspackman I've chosen to implement the solution from comment  https://github.com/qooxdoo/qooxdoo/pull/9499#discussion_r170922372 if we can live with that.

This fixes the bug with the `arguments.callee` way. If `arguments.callee` is not present the old way sort functions will break anyway be throwing errors on `arguments.callee.columnIndex` access.